### PR TITLE
HTTP::Request, added remote ip and peer_addr

### PR DESCRIPTION
**Pull request by benoist**

### CODE REVIEW

File: src/http/request.cr [hunk 1]
```diff
@@ -33,6 +33,27 @@ class HTTP::Request
     @method == "HEAD"
   end
 
+  def peer_addr=(addr)
+    @peer_addr = addr
+  end
+
+  def peer_addr
+    @peer_addr.not_nil!
+  end
```
Comment (jhass) [inline src/http/request.cr]: This is `property! peer_addr`

File: src/http/request.cr [hunk 2]
```diff
@@ -68,4 +89,21 @@ class HTTP::Request
   private def uri
     (@uri ||= URI.parse(@resource)).not_nil!
   end
+
+  private def ips_from(header)
+    if ips = headers[header]? || headers["Http-#{header.tr("_", "-")}"]?
```
Comment (jhass) [inline src/http/request.cr]: `HTTP::Headers` should have independent lookup by now, https://github.com/manastech/crystal/blob/master/spec/std/http/headers_spec.cr

File: src/http/request.cr (hunk 1 above)
Comment (benoist) [inline src/http/request.cr]: Good to know :)

File: src/http/request.cr (hunk 2 above)
Comment (benoist) [inline src/http/request.cr]: But it can't lookup HTTP_X_FORWARDED_FOR and X-Forwarded-For right?

Comment (benoist) [inline src/http/request.cr]: So only the tr("_", "-") is redundant right?

File: src/http/request.cr [hunk 3]
```diff
@@ -68,4 +89,21 @@ class HTTP::Request
...
+      [] of String
+    end
+  end
+
+  private def first_remote_ip_address(ip_addresses)
+    ip_addresses.find { |ip| !trusted_proxy?(ip) }
+  end
+
+  private def trusted_proxy?(ip)
+    ip =~ /\A127\.0\.0\.1\Z|\A(10|172\.(1[6-9]|2[0-9]|30|31)|192\.168)\.|\A::1\Z|\Afd[0-9a-f]{2}:.+|\Alocalhost\Z|\Aunix\Z|\Aunix:/i
```
Comment (jhass) [inline src/http/request.cr]: *  `127.0.0.0/8` https://tools.ietf.org/html/rfc6890
* Should we trust link-local addresses? `fe80::/10`

I also wonder if given that we shouldn't have `remote_ip(trust_headers=true, trust_all_clients=false)`?

File: src/http/request.cr [hunk 4]
```diff
@@ -33,6 +34,19 @@ class HTTP::Request
     @method == "HEAD"
   end
 
+  def remote_ip(trust_headers=true)
+    if trust_headers
+      client_ips = ips_from("Client-Ip").reverse
+      forwarded_ips = ips_from("X-Forwarded-For").reverse
```
Comment (jhass) [inline src/http/request.cr]: As said I've also seen `"X-Real-Ip"` in configs and `X-Fowarded`, `Forwarded` and `"X-Cluster-Client-Ip"` appear to be used values too.

File: src/http/request.cr [hunk 5]
```diff
@@ -68,4 +82,21 @@ class HTTP::Request
   private def uri
     (@uri ||= URI.parse(@resource)).not_nil!
   end
+
+  private def ips_from(header)
+    if ips = headers[header]? || headers["Http-#{header}"]?
+      ips.strip.split(/[,\s]+/)
+    else
+      [] of String
```
Comment (jhass) [inline src/http/request.cr]: Mhh, `Tuple.new` should work too here, no?

File: src/http/request.cr [hunk 6]
```diff
@@ -33,6 +34,19 @@ class HTTP::Request
     @method == "HEAD"
   end
 
+  def remote_ip(trust_headers=true)
+    if trust_headers
+      client_ips = ips_from("Client-Ip").reverse
+      forwarded_ips = ips_from("X-Forwarded-For").reverse
+
+      ips = [forwarded_ips, client_ips, peer_addr.ip_address].flatten.compact
```
Comment (jhass) [inline src/http/request.cr]: Mhh, `Array#+` perhaps?

File: src/http/request.cr (hunk 2 above)
Comment (jhass) [inline src/http/request.cr]: Eh, missed the comment,  sorry. But yes, you're right and looks good now.

File: spec/std/http/request_spec.cr [hunk 7]
```diff
@@ -203,5 +204,69 @@ module HTTP
...
+      end
+    end
+
+    describe "#peer_addr" do
+      it "returns the peer_addr" do
+        request = Request.from_io(StringIO.new("GET / HTTP/1.1\r\nHost: host.example.org\r\n\r\n")).not_nil!
+        addr = Socket::Addr.new("AF_INET", "12345", "127.0.0.1")
+        request.peer_addr = addr
+        request.peer_addr.should eq addr
+      end
```
Comment (lbguilherme) [inline spec/std/http/request_spec.cr]: What's the point of having two identical specs?

**Comment by waj** [comment]
I'm not so sure about this. I think we're adding too much intelligence to the Request already. Maybe we need to put this logic in a handler that we insert as a middleware. 

Also, note that the same class is used for client requests so I wonder if we shouldn't have a context object instead that we pass around between server handlers.

**Comment by benoist** [comment]
@waj that was my first suggestion. I only added the peeraddr to the REMOTE_ADDR header to be used in middleware. I like the peer_addr on request better. But perhaps we can move the remote ip logic to a handler that sets the REMOTE_ADDR header? Or is there another way of passing that info down the middleware stack without using the headers?

**Comment by jhass** [comment]
I would really dislike adding custom headers. And a middleware approach would do the detection on every single request, keep in mind it's about 10 header checks per request.

**Comment by benoist** [comment]
But if you use it for logging purposes you're gonna be doing the lookups on every request either via the middleware or via the request object.
If you only need the remote ip let's say every ten requests then it's 10 headers checks you're not gonna need

But there's always the option of creating an extra class where the logic can be placed. We could also create an HTTP util class.

But in all cases do we agree to the peer_addr being set on the request? Cause I would happily split the pull request into peer_addr and remote_ip

**Comment by waterlink** [comment]
How about having a middleware that calls underlying handlers with request decorator, that has all the additional logic added on top of raw request? i.e.:

```crystal
class SmartHTTPHandler < HTTP::Handler
  def call(request)
    smart_request = SmartRequest.new(request)
    call_next(smart_request)
  end
end

class SmartRequest
  private getter original_request
  def initialize(@original_request)
  end

  delegate_when_missing_to original_request

  # lazily fetches peer address of request
  def peer_addr
    # ...
  end

  # set peer address of request
  def peer_addr=(value)
    # ...
  end

  # .. and more stuff, like #url_params, #body_params, etc. ..
end
```

**Comment by waterlink** [comment]
Other point, is that middleware is not required in this case, one can use these decorators directly on demand.

**Comment by asterite** [comment]
We plan to make a review of http request/response and client with @waj for this, and we'll hopefully come with a good solution.

**Comment by sdogruyol** [comment]
@asterite now that https://github.com/crystal-lang/crystal/pull/3406 is merged WDYT about this?

**Comment by asterite** [comment]
@sdogruyol It's on our TODO list, just that we won't implement it this way (it doesn't have to be attached to the request, it will probably be in the HTTP::Server::Context).

**Comment by TheLonelyGhost** [comment]
I originally came here to make sure this doesn't consider the HTTP headers as the authoritative source for IP address since HTTP headers can be spoofed. Glad to see there is an option to rely on the socket information (coming from the TCP connection data itself) or the HTTP header.

As far as code quality is concerned, LGTM. Perhaps some cleanup of the specs according to the comments already left by others.

LGTM :+1:

**Comment by rdp** [comment]
FWIW it would be "convenient" to have this in there I think people would use it.  Here's how rails 2.3 does it: https://gist.github.com/rdp/ec8d8d7152988f496f951a2f0d65a320

**Comment by oltmannsdaniel** [comment]
Still not merged!! I really need that right now 😅   @benoist

**Comment by benoist** [comment]
Well i think this code is pretty outdated. Happy to make the required changes to get this merged though.

File: src/http/request.cr (hunk 4 above)
Comment (rdp) [inline src/http/request.cr]: See also https://github.com/jhass/carc.in/blob/master/src/web.cr#L96-L102 FWIW

**Comment by rdp** [comment]
@benoist Yeah I'd say do it, if not I will in a couple weeks I think...

**Comment by sdogruyol** [comment]
Wow, i still remember the day this issue opened. Time really flies 😄

**Comment by benoist** [comment]
@rdp Do we know what the required changes are, as I believe @asterite was going to discuss the implementation details with @waj

**Comment by rdp** [comment]
I assume just the review comments but everybody feel free to chime in :)

**Comment by benoist** [comment]
> @sdogruyol It's on our TODO list, just that we won't implement it this way (it doesn't have to be attached to the request, it will probably be in the HTTP::Server::Context).

That was after the review comments from jhass

**Comment by asterite** [comment]
This is too old so I'm closing it.

We still need this functionality, though, but it will probably come from the core team. Maybe we can copy [what Go does](https://stackoverflow.com/a/27235620), but I don't know (I don't know much about HTTP and proxies).

**Pull request closed without merging**